File: tests/oracle/typeop_localbase_defaults_1204.cc
/*
 * TYPEOP-LOCALBASE-DEFAULTS-0001: locked Ghidra 12.0.4 oracle projection for
 * the TypeOp base-class getOutputLocal/getInputLocal defaults and the
 * TypeOpCall constructor opflags.
 *
 * Every observation goes through PcodeOp::inputTypeLocal/outputTypeLocal, so
 * the Architecture-owned TypeOp table instance is the code under
 * observation. Cases cover:
 *   - opcodes with no get*Local override (BRANCH/BRANCHIND/SEGMENTOP/CAST)
 *     whose local type must be getBase(size,TYPE_UNKNOWN),
 *   - the TypeOpFunc metain/metaout derivation recorded through ZEXT (UINT)
 *     and SEXT (INT),
 *   - the CALL op without a FSPEC annotation (base defaults through
 *     TypeOpCall's explicit TypeOp::get*Local fallbacks) and with a locked
 *     parameter (D1 regression),
 *   - non-standard operand sizes (3 bytes) through TypeFactory::getBase,
 *   - the TypeOpCall constructor opflags bit for bit.
 */
#include "bfd_arch.hh"
#include "funcdata.hh"
#include "libdecomp.hh"
#include "typeop.hh"

#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

using namespace ghidra;
using namespace std;

string typeToken(const Datatype *ct)

{
  ostringstream stream;
  ct->printRaw(stream);
  return stream.str();
}

void emitBool(const string &key,bool value)

{
  cout << key << '=' << (value ? 1 : 0) << '\n';
}

// One input-local observation: inputTypeLocal(slot) must be
// getBase(op->getIn(slot)->getSize(),TYPE_UNKNOWN) for base-default opcodes.
void emitInputCase(const string &name,PcodeOp *op,int4 slot,TypeFactory *factory)

{
  Datatype *actual = op->inputTypeLocal(slot);
  Datatype *repeat = op->inputTypeLocal(slot);
  Datatype *fallback = factory->getBase(op->getIn(slot)->getSize(),TYPE_UNKNOWN);
  cout << "case." << name << ".slot=" << slot << '\n';
  cout << "case." << name << ".operand_size=" << op->getIn(slot)->getSize() << '\n';
  cout << "case." << name << ".result_type=" << typeToken(actual) << '\n';
  cout << "case." << name << ".result_meta=" << static_cast<int4>(actual->getMetatype()) << '\n';
  cout << "case." << name << ".result_size=" << actual->getSize() << '\n';
  emitBool("case." + name + ".base_identity",actual == fallback);
  emitBool("case." + name + ".repeat_identity",actual == repeat);
}

// One output-local observation: outputTypeLocal() must be
// getBase(op->getOut()->getSize(),TYPE_UNKNOWN) for base-default opcodes.
void emitOutputCase(const string &name,PcodeOp *op,TypeFactory *factory)

{
  Datatype *actual = op->outputTypeLocal();
  Datatype *repeat = op->outputTypeLocal();
  Datatype *fallback = factory->getBase(op->getOut()->getSize(),TYPE_UNKNOWN);
  cout << "case." << name << ".slot=-1" << '\n';
  cout << "case." << name << ".operand_size=" << op->getOut()->getSize() << '\n';
  cout << "case." << name << ".result_type=" << typeToken(actual) << '\n';
  cout << "case." << name << ".result_meta=" << static_cast<int4>(actual->getMetatype()) << '\n';
  cout << "case." << name << ".result_size=" << actual->getSize() << '\n';
  emitBool("case." + name + ".base_identity",actual == fallback);
  emitBool("case." + name + ".repeat_identity",actual == repeat);
}

// ZEXT/SEXT record the TypeOpFunc metain/metaout derivation: the input local
// is getBase(inSize,metain) and the output local getBase(outSize,metaout),
// with metain/metaout=TYPE_UINT for ZEXT and TYPE_INT for SEXT (typeop.cc
// :1116/:1142 constructors; selectJavaOperators can retune them).
void emitMetatypeCase(const string &name,PcodeOp *op,int4 slot,type_metatype meta,
                      TypeFactory *factory)

{
  Datatype *actual = (slot < 0) ? op->outputTypeLocal() : op->inputTypeLocal(slot);
  int4 size = (slot < 0) ? op->getOut()->getSize() : op->getIn(slot)->getSize();
  Datatype *metaBase = factory->getBase(size,meta);
  Datatype *unknownBase = factory->getBase(size,TYPE_UNKNOWN);
  cout << "case." << name << ".slot=" << slot << '\n';
  cout << "case." << name << ".operand_size=" << size << '\n';
  cout << "case." << name << ".result_type=" << typeToken(actual) << '\n';
  cout << "case." << name << ".result_meta=" << static_cast<int4>(actual->getMetatype()) << '\n';
  cout << "case." << name << ".result_size=" << actual->getSize() << '\n';
  emitBool("case." + name + ".meta_identity",actual == metaBase);
  emitBool("case." + name + ".base_identity",actual == unknownBase);
}

ParameterPieces pieces(AddrSpace *space,uintb offset,Datatype *type,uint4 flags)

{
  ParameterPieces result;
  result.addr = Address(space,offset);
  result.type = type;
  result.flags = flags;
  return result;
}

void runFixture(const string &specDirectory,const string &binary)

{
  vector<string> specPaths(1,specDirectory);
  startDecompilerLibrary(specPaths);
  BfdArchitecture architecture(binary,"default",&cerr);
  DocumentStorage store;
  architecture.init(store);
  if (architecture.archid != "x86:LE:64:default:gcc")
    throw runtime_error("runtime architecture/compiler drifted: " + architecture.archid);

  TypeFactory *factory = architecture.types;
  AddrSpace *code = architecture.getDefaultCodeSpace();
  AddrSpace *reg = architecture.getSpaceByName("register");
  if (factory == (TypeFactory *)0 || code == (AddrSpace *)0 || reg == (AddrSpace *)0)
    throw runtime_error("required architecture service missing");

  Scope *global = architecture.symboltab->getGlobalScope();
  FunctionSymbol *symbol = global->addFunction(
      Address(code,0x500000),"typeop_localbase_fixture");
  Funcdata *fd = symbol->getFunction();

  cout << "fixture=TYPEOP-LOCALBASE-DEFAULTS-0001" << '\n';
  cout << "architecture=" << architecture.archid << '\n';

  // ---- TypeOpCall constructor opflags (typeop.cc:663) ----
  TypeOpCall callInst(factory);
  const uint4 callFlags = callInst.getFlags();
  {
    ostringstream stream;
    stream << hex << callFlags;
    cout << "flags.call.hex=" << stream.str() << '\n';
  }
  cout << "flags.call.decimal=" << callFlags << '\n';
  emitBool("flags.call.special",(callFlags & PcodeOp::special) != 0);
  emitBool("flags.call.call_bit",(callFlags & PcodeOp::call) != 0);
  emitBool("flags.call.has_callspec",(callFlags & PcodeOp::has_callspec) != 0);
  emitBool("flags.call.coderef",(callFlags & PcodeOp::coderef) != 0);
  emitBool("flags.call.nocollapse",(callFlags & PcodeOp::nocollapse) != 0);
  emitBool("flags.call.commutative_clear",(callFlags & PcodeOp::commutative) == 0);

  // ---- BRANCH: base defaults on the coderef input (no output) ----
  PcodeOp *branchOp = fd->newOp(1,Address(code,0x500100));
  fd->opSetOpcode(branchOp,CPUI_BRANCH);
  fd->opSetInput(branchOp,fd->newCodeRef(Address(code,0x600100)),0);
  emitInputCase("branch_input0",branchOp,0,factory);

  // ---- BRANCHIND: base defaults on a 4-byte constant input ----
  PcodeOp *branchindOp = fd->newOp(1,Address(code,0x500110));
  fd->opSetOpcode(branchindOp,CPUI_BRANCHIND);
  fd->opSetInput(branchindOp,fd->newConstant(4,0x1234),0);
  emitInputCase("branchind_input0",branchindOp,0,factory);

  // ---- SEGMENTOP: base defaults incl. a non-standard 3-byte operand ----
  PcodeOp *segOp = fd->newOp(3,Address(code,0x500120));
  fd->opSetOpcode(segOp,CPUI_SEGMENTOP);
  fd->opSetInput(segOp,fd->newConstant(4,0),0);
  fd->opSetInput(segOp,fd->newConstant(3,0x112233),1);
  fd->opSetInput(segOp,fd->newConstant(8,0),2);
  fd->opSetOutput(segOp,fd->newVarnode(3,reg,0x300));
  emitInputCase("segment_input1_size3",segOp,1,factory);
  emitOutputCase("segment_output_size3",segOp,factory);

  // ---- CAST: "we don't care what types are cast" -> base defaults ----
  PcodeOp *castOp = fd->newOp(1,Address(code,0x500130));
  fd->opSetOpcode(castOp,CPUI_CAST);
  fd->opSetInput(castOp,fd->newConstant(4,0xaabbccdd),0);
  fd->opSetOutput(castOp,fd->newVarnode(4,reg,0x400));
  emitInputCase("cast_input0",castOp,0,factory);
  emitOutputCase("cast_output",castOp,factory);

  // ---- ZEXT/SEXT: TypeOpFunc metain/metaout derivation (UINT vs INT) ----
  PcodeOp *zextOp = fd->newOp(1,Address(code,0x500140));
  fd->opSetOpcode(zextOp,CPUI_INT_ZEXT);
  fd->opSetInput(zextOp,fd->newConstant(1,0x7f),0);
  fd->opSetOutput(zextOp,fd->newVarnode(4,reg,0x500));
  emitMetatypeCase("zext_input0",zextOp,0,TYPE_UINT,factory);
  emitMetatypeCase("zext_output",zextOp,-1,TYPE_UINT,factory);

  PcodeOp *sextOp = fd->newOp(1,Address(code,0x500150));
  fd->opSetOpcode(sextOp,CPUI_INT_SEXT);
  fd->opSetInput(sextOp,fd->newConstant(1,0x80),0);
  fd->opSetOutput(sextOp,fd->newVarnode(4,reg,0x510));
  emitMetatypeCase("sext_input0",sextOp,0,TYPE_INT,factory);
  emitMetatypeCase("sext_output",sextOp,-1,TYPE_INT,factory);

  // ---- CALL without a FSPEC annotation: TypeOpCall explicit fallbacks ----
  // resolve through the TypeOp base defaults (typeop.cc:696/:729).
  const int4 pointerSize = factory->getSizeOfPointer();
  const int4 wordSize = architecture.getDefaultDataSpace()->getWordSize();
  Datatype *voidType = factory->getTypeVoid();
  Datatype *charType = factory->getTypeChar(factory->getSizeOfChar());
  Datatype *charPointer = factory->getTypePointer(pointerSize,charType,wordSize);

  PcodeOp *plainCallOp = fd->newOp(1,Address(code,0x500160));
  fd->opSetOpcode(plainCallOp,CPUI_CALL);
  // A direct-call target before flow analysis: input 0 is still the coderef.
  fd->opSetInput(plainCallOp,fd->newCodeRef(Address(code,0x600160)),0);
  fd->opSetOutput(plainCallOp,fd->newVarnode(8,reg,0x600));
  emitInputCase("call_input0_nofspec",plainCallOp,0,factory);
  emitOutputCase("call_output_nofspec",plainCallOp,factory);

  // ---- CALL with one locked parameter: D1 behavior must not regress ----
  PcodeOp *callOp = fd->newOp(2,Address(code,0x500170));
  fd->opSetOpcode(callOp,CPUI_CALL);
  fd->opSetInput(callOp,fd->newCodeRef(Address(code,0x600170)),0);
  FuncCallSpecs callspec(callOp);
  callspec.setInternal(architecture.defaultfp,voidType);
  callspec.setParam(0,"locked_ptr",
      pieces(reg,0x00,charPointer,ParameterPieces::typelock));
  fd->opSetInput(callOp,fd->newVarnodeCallSpecs(&callspec),0);
  fd->opSetInput(callOp,fd->newConstant(8,0x7180),1);
  {
    Datatype *actual = callOp->inputTypeLocal(1);
    Datatype *repeat = callOp->inputTypeLocal(1);
    Datatype *fallback = factory->getBase(8,TYPE_UNKNOWN);
    cout << "case.call_input1_locked.slot=1" << '\n';
    cout << "case.call_input1_locked.operand_size=8" << '\n';
    cout << "case.call_input1_locked.result_type=" << typeToken(actual) << '\n';
    cout << "case.call_input1_locked.result_meta="
         << static_cast<int4>(actual->getMetatype()) << '\n';
    cout << "case.call_input1_locked.result_size=" << actual->getSize() << '\n';
    emitBool("case.call_input1_locked.param_identity",actual == charPointer);
    emitBool("case.call_input1_locked.base_identity",actual == fallback);
    emitBool("case.call_input1_locked.repeat_identity",actual == repeat);
  }
}

} // namespace

int main(int argc,char **argv)

{
  try {
    if (argc != 3)
      throw invalid_argument("usage: typeop_localbase_defaults_1204 SPEC_DIRECTORY BINARY");
    runFixture(argv[1],argv[2]);
    return 0;
  }
  catch(const LowlevelError &err) {
    cerr << "LowlevelError: " << err.explain << '\n';
  }
  catch(const exception &err) {
    cerr << "std::exception: " << err.what() << '\n';
  }
  return 1;
}

File: tests/oracle/typeop_localbase_defaults_1204.rs
// TYPEOP-LOCALBASE-DEFAULTS-0001: current-Rugra comparand for the locked
// Ghidra TypeOp base-class getOutputLocal/getInputLocal defaults and the
// TypeOpCall constructor opflags.
//
// Every observation goes through the production TypeOpManager dispatch table
// built with the same TypeFactory allocation the fixture Architecture uses,
// so the base-default cases exercise the shared `base_local_type` helper
// through the `local_type_factory` provider hook exactly as production would.

use std::sync::{Arc, RwLock};

use rugra::address::Address;
use rugra::arch::Architecture;
use rugra::fspec::{protoparam_flags, FuncCallSpecs, FuncProto, ProtoParameter};
use rugra::funcdata::Funcdata;
use rugra::marshal::{Element, IdRegistry, TreeDecoder};
use rugra::op::PcodeOp;
use rugra::opcodes::OpCode;
use rugra::type_system::datatype::{Datatype, TypeMetatype};
use rugra::type_system::typefactory::{SizeArchInputs, TypeFactory};
use rugra::typeop::{TypeOp, TypeOpManager};

fn element(name: &str, attributes: &[(&str, &str)]) -> Arc<RwLock<Element>> {
    let mut element = Element::new();
    element.set_name(name);
    for (key, value) in attributes {
        element.add_attribute(key, value);
    }
    Arc::new(RwLock::new(element))
}

/// Mirror of the locked BfdArchitecture TypeFactory state the C++ fixture
/// observes (same bootstrap as tests/oracle/typeop_local_type_1204.rs):
/// `TypeFactory::raw()`, the x86-64-gcc.cspec `<size_alignment_map>`
/// (entries 1,2,4,8,16), `setup_sizes` defaults for a 64-bit stack pointer,
/// and the `SleighArchitecture::buildCoreTypes` default core set.
fn configure_factory() -> TypeFactory {
    let mut factory = TypeFactory::raw();

    let alignment_map = element("size_alignment_map", &[]);
    for (size, alignment) in [("1", "1"), ("2", "2"), ("4", "4"), ("8", "8"), ("16", "16")] {
        alignment_map
            .write()
            .unwrap()
            .add_child(element("entry", &[("size", size), ("alignment", alignment)]));
    }
    let organization = element("data_organization", &[]);
    organization.write().unwrap().add_child(alignment_map);
    let registry = Arc::new(RwLock::new(IdRegistry::new()));
    let mut organization_decoder = TreeDecoder::new(organization, registry.clone());
    factory.decode_data_organization(&mut organization_decoder);

    // sleigh_arch.cc:215-236 order preserved.
    let core_types: &[(&str, usize, TypeMetatype, bool)] = &[
        ("void", 1, TypeMetatype::Void, false),
        ("bool", 1, TypeMetatype::Bool, false),
        ("uint1", 1, TypeMetatype::Uint, false),
        ("uint2", 2, TypeMetatype::Uint, false),
        ("uint4", 4, TypeMetatype::Uint, false),
        ("uint8", 8, TypeMetatype::Uint, false),
        ("int1", 1, TypeMetatype::Int, false),
        ("int2", 2, TypeMetatype::Int, false),
        ("int4", 4, TypeMetatype::Int, false),
        ("int8", 8, TypeMetatype::Int, false),
        ("float4", 4, TypeMetatype::Float, false),
        ("float8", 8, TypeMetatype::Float, false),
        ("float10", 10, TypeMetatype::Float, false),
        ("float16", 16, TypeMetatype::Float, false),
        ("xunknown1", 1, TypeMetatype::Unknown, false),
        ("xunknown2", 2, TypeMetatype::Unknown, false),
        ("xunknown4", 4, TypeMetatype::Unknown, false),
        ("xunknown8", 8, TypeMetatype::Unknown, false),
        ("code", 1, TypeMetatype::Code, false),
        ("char", 1, TypeMetatype::Int, true),
        ("wchar2", 2, TypeMetatype::Int, true),
        ("wchar4", 4, TypeMetatype::Int, true),
    ];
    for (name, size, meta, chartp) in core_types {
        factory
            .set_core_type_result(name, *size, *meta, *chartp)
            .unwrap_or_else(|message| panic!("core registration {name}: {message}"));
    }
    factory.cache_core_types();

    factory.setup_sizes(&SizeArchInputs {
        stack_spacebase_size: Some(8),
        default_data_space_addr_size: 8,
        default_size: 8,
        far_pointer: None,
    });
    factory
}

fn emit_bool(key: &str, value: bool) {
    println!("{key}={}", if value { 1 } else { 0 });
}

fn factory_base(
    factory: &Arc<RwLock<TypeFactory>>,
    size: usize,
    metatype: TypeMetatype,
) -> Arc<Datatype> {
    factory
        .read()
        .unwrap()
        .get_base(size, metatype)
        .expect("fixture canonical base type")
}

fn ghidra_metatype(metatype: TypeMetatype) -> i32 {
    match metatype {
        TypeMetatype::PartialUnion => 0,
        TypeMetatype::PartialStruct => 1,
        TypeMetatype::PartialEnum => 2,
        TypeMetatype::Union => 3,
        TypeMetatype::Struct => 4,
        TypeMetatype::Enum => 5,
        TypeMetatype::Array => 7,
        TypeMetatype::Pointer => 9,
        TypeMetatype::Float => 10,
        TypeMetatype::Code => 11,
        TypeMetatype::Bool => 12,
        TypeMetatype::Uint => 13,
        TypeMetatype::Int => 14,
        TypeMetatype::Unknown => 15,
        TypeMetatype::Spacebase => 16,
        TypeMetatype::Void => 17,
    }
}

/// Byte projection of `Datatype::printRaw` for the bounded result set these
/// cases can return (TypeBase and TypePointer): type.cc:139-146 prints the
/// name or `unkbyte<size>` for every non-pointer type, and type.cc:910-916
/// appends `" *"` for pointers (factory pointer products carry no spaceid,
/// so there is no suffix beyond that).
fn ghidra_print_raw(datatype: &Datatype) -> String {
    if let Datatype::Pointer(pointer) = datatype {
        return format!("{} *", ghidra_print_raw(&pointer.ptr_to));
    }
    let name = datatype.get_name().to_string();
    if !name.is_empty() {
        name
    } else {
        format!("unkbyte{}", datatype.get_size())
    }
}

fn emit_result(name: &str, slot: i32, size: usize, actual: &Option<Arc<Datatype>>) {
    println!("case.{name}.slot={slot}");
    println!("case.{name}.operand_size={size}");
    match actual.as_ref() {
        Some(datatype) => {
            println!("case.{name}.result_type={}", ghidra_print_raw(datatype));
            println!(
                "case.{name}.result_meta={}",
                ghidra_metatype(datatype.get_metatype())
            );
            println!("case.{name}.result_size={}", datatype.get_size());
        }
        None => {
            println!("case.{name}.result_type=NONE");
            println!("case.{name}.result_meta=NONE");
            println!("case.{name}.result_size=NONE");
        }
    }
}

fn identity(actual: &Option<Arc<Datatype>>, expected: &Arc<Datatype>) -> bool {
    actual.as_ref().is_some_and(|datatype| Arc::ptr_eq(datatype, expected))
}

/// Input-local observation through the production dispatch table; the
/// expected fallback is getBase(op.get_in(slot).size, UNKNOWN).
fn emit_input_case(
    name: &str,
    typeop: &dyn TypeOp,
    op: &PcodeOp,
    slot: usize,
    type_factory: &Arc<RwLock<TypeFactory>>,
) {
    let size = op
        .get_in(slot)
        .expect("fixture input slot")
        .read()
        .unwrap()
        .get_size();
    let actual = typeop.get_input_local(op, slot);
    let repeat = typeop.get_input_local(op, slot);
    emit_result(name, slot as i32, size, &actual);
    let fallback = factory_base(type_factory, size, TypeMetatype::Unknown);
    emit_bool(
        &format!("case.{name}.base_identity"),
        identity(&actual, &fallback),
    );
    emit_bool(
        &format!("case.{name}.repeat_identity"),
        match (actual.as_ref(), repeat.as_ref()) {
            (Some(left), Some(right)) => Arc::ptr_eq(left, right),
            _ => false,
        },
    );
}

/// Output-local observation through the production dispatch table; the
/// expected fallback is getBase(op.get_out().size, UNKNOWN).
fn emit_output_case(
    name: &str,
    typeop: &dyn TypeOp,
    op: &PcodeOp,
    type_factory: &Arc<RwLock<TypeFactory>>,
) {
    let size = op
        .get_out()
        .expect("fixture output varnode")
        .read()
        .unwrap()
        .get_size();
    let actual = typeop.get_output_local(op);
    let repeat = typeop.get_output_local(op);
    emit_result(name, -1, size, &actual);
    let fallback = factory_base(type_factory, size, TypeMetatype::Unknown);
    emit_bool(
        &format!("case.{name}.base_identity"),
        identity(&actual, &fallback),
    );
    emit_bool(
        &format!("case.{name}.repeat_identity"),
        match (actual.as_ref(), repeat.as_ref()) {
            (Some(left), Some(right)) => Arc::ptr_eq(left, right),
            _ => false,
        },
    );
}

/// ZEXT/SEXT metatype-derivation observation. The locked oracle derives the
/// local type as getBase(size, metain/metaout) with UINT (ZEXT) / INT (SEXT);
/// the current Rugra macro override reads the opposite varnode's v_type, so
/// an untyped fixture op surfaces the PRINTC-CAST-OPNAME-0001 M1 residual.
fn emit_metatype_case(
    name: &str,
    typeop: &dyn TypeOp,
    op: &PcodeOp,
    slot: i32,
    meta: TypeMetatype,
    type_factory: &Arc<RwLock<TypeFactory>>,
) {
    let (actual, size) = if slot < 0 {
        let size = op
            .get_out()
            .expect("fixture output varnode")
            .read()
            .unwrap()
            .get_size();
        (typeop.get_output_local(op), size)
    } else {
        let size = op
            .get_in(slot as usize)
            .expect("fixture input slot")
            .read()
            .unwrap()
            .get_size();
        (typeop.get_input_local(op, slot as usize), size)
    };
    emit_result(name, slot, size, &actual);
    let meta_base = factory_base(type_factory, size, meta);
    let unknown_base = factory_base(type_factory, size, TypeMetatype::Unknown);
    emit_bool(
        &format!("case.{name}.meta_identity"),
        identity(&actual, &meta_base),
    );
    emit_bool(
        &format!("case.{name}.base_identity"),
        identity(&actual, &unknown_base),
    );
}

fn main() {
    let mut factory = configure_factory();

    let void_type = factory.get_type_void();
    let char_type = factory
        .get_type_char_named("char")
        .expect("canonical char type");
    let char_pointer = factory.get_type_pointer(8, char_type, 1);

    let type_factory = Arc::new(RwLock::new(factory));
    let mut architecture = Architecture::new();
    architecture.archid = "x86:LE:64:default:gcc".to_string();
    architecture.set_types(type_factory.clone());

    let mut fd = Funcdata::new("typeop_localbase_fixture", Address::new(0x500000), 0x100);
    fd.vbank.set_type_factory(type_factory.clone());

    // Production dispatch table built exactly like TypeOpManager::new does.
    let manager = TypeOpManager::new(type_factory.clone());

    println!("fixture=TYPEOP-LOCALBASE-DEFAULTS-0001");
    println!("architecture={}", architecture.archid);

    // ---- TypeOpCall constructor opflags (typeop.cc:663) ----
    let call_typeop = manager
        .get_op(OpCode::CPUI_CALL)
        .expect("registered CALL TypeOp");
    let call_flags = call_typeop.get_flags();
    println!("flags.call.hex={:x}", call_flags);
    println!("flags.call.decimal={}", call_flags);
    emit_bool(
        "flags.call.special",
        call_flags & rugra::op::pcodeop_flags::SPECIAL != 0,
    );
    emit_bool(
        "flags.call.call_bit",
        call_flags & rugra::op::pcodeop_flags::CALL != 0,
    );
    emit_bool(
        "flags.call.has_callspec",
        call_flags & rugra::op::pcodeop_flags::HAS_CALLSPEC != 0,
    );
    emit_bool(
        "flags.call.coderef",
        call_flags & rugra::op::pcodeop_flags::CODEREF != 0,
    );
    emit_bool(
        "flags.call.nocollapse",
        call_flags & rugra::op::pcodeop_flags::NOCOLLAPSE != 0,
    );
    emit_bool(
        "flags.call.commutative_clear",
        call_flags & rugra::op::pcodeop_flags::COMMUTATIVE == 0,
    );

    // ---- BRANCH: base defaults on the coderef input ----
    {
        let branch_typeop = manager
            .get_op(OpCode::CPUI_BRANCH)
            .expect("registered BRANCH TypeOp");
        let target = fd.new_code_ref(Address::new(0x600100));
        let op = fd.new_op(1, Address::new(0x500100));
        fd.op_set_opcode(&op, OpCode::CPUI_BRANCH);
        fd.op_set_input(&op, target, 0);
        let call = op.0.read().unwrap();
        emit_input_case("branch_input0", branch_typeop, &call, 0, &type_factory);
    }

    // ---- BRANCHIND: base defaults on a 4-byte constant input ----
    {
        let branchind_typeop = manager
            .get_op(OpCode::CPUI_BRANCHIND)
            .expect("registered BRANCHIND TypeOp");
        let target = fd.new_constant(4, 0x1234);
        let op = fd.new_op(1, Address::new(0x500110));
        fd.op_set_opcode(&op, OpCode::CPUI_BRANCHIND);
        fd.op_set_input(&op, target, 0);
        let call = op.0.read().unwrap();
        emit_input_case("branchind_input0", branchind_typeop, &call, 0, &type_factory);
    }

    // ---- SEGMENTOP: base defaults incl. a non-standard 3-byte operand ----
    {
        let segment_typeop = manager
            .get_op(OpCode::CPUI_SEGMENTOP)
            .expect("registered SEGMENTOP TypeOp");
        let spaceid = fd.new_constant(4, 0);
        let offset3 = fd.new_constant(3, 0x112233);
        let base = fd.new_constant(8, 0);
        let out = fd.new_varnode(3, Address::new(0x300));
        let op = fd.new_op(3, Address::new(0x500120));
        fd.op_set_opcode(&op, OpCode::CPUI_SEGMENTOP);
        fd.op_set_input(&op, spaceid, 0);
        fd.op_set_input(&op, offset3, 1);
        fd.op_set_input(&op, base, 2);
        fd.op_set_output(&op, out);
        let call = op.0.read().unwrap();
        emit_input_case(
            "segment_input1_size3",
            segment_typeop,
            &call,
            1,
            &type_factory,
        );
        emit_output_case("segment_output_size3", segment_typeop, &call, &type_factory);
    }

    // ---- CAST: "we don't care what types are cast" -> base defaults ----
    {
        let cast_typeop = manager
            .get_op(OpCode::CPUI_CAST)
            .expect("registered CAST TypeOp");
        let input0 = fd.new_constant(4, 0xaabbccdd);
        let out = fd.new_varnode(4, Address::new(0x400));
        let op = fd.new_op(1, Address::new(0x500130));
        fd.op_set_opcode(&op, OpCode::CPUI_CAST);
        fd.op_set_input(&op, input0, 0);
        fd.op_set_output(&op, out);
        let call = op.0.read().unwrap();
        emit_input_case("cast_input0", cast_typeop, &call, 0, &type_factory);
        emit_output_case("cast_output", cast_typeop, &call, &type_factory);
    }

    // ---- ZEXT/SEXT: TypeOpFunc metain/metaout derivation (UINT vs INT) ----
    {
        let zext_typeop = manager
            .get_op(OpCode::CPUI_INT_ZEXT)
            .expect("registered ZEXT TypeOp");
        let input0 = fd.new_constant(1, 0x7f);
        let out = fd.new_varnode(4, Address::new(0x500));
        let op = fd.new_op(1, Address::new(0x500140));
        fd.op_set_opcode(&op, OpCode::CPUI_INT_ZEXT);
        fd.op_set_input(&op, input0, 0);
        fd.op_set_output(&op, out);
        let call = op.0.read().unwrap();
        emit_metatype_case("zext_input0", zext_typeop, &call, 0, TypeMetatype::Uint, &type_factory);
        emit_metatype_case("zext_output", zext_typeop, &call, -1, TypeMetatype::Uint, &type_factory);
    }
    {
        let sext_typeop = manager
            .get_op(OpCode::CPUI_INT_SEXT)
            .expect("registered SEXT TypeOp");
        let input0 = fd.new_constant(1, 0x80);
        let out = fd.new_varnode(4, Address::new(0x510));
        let op = fd.new_op(1, Address::new(0x500150));
        fd.op_set_opcode(&op, OpCode::CPUI_INT_SEXT);
        fd.op_set_input(&op, input0, 0);
        fd.op_set_output(&op, out);
        let call = op.0.read().unwrap();
        emit_metatype_case("sext_input0", sext_typeop, &call, 0, TypeMetatype::Int, &type_factory);
        emit_metatype_case("sext_output", sext_typeop, &call, -1, TypeMetatype::Int, &type_factory);
    }

    // ---- CALL without a FSPEC annotation: TypeOpCall explicit fallbacks
    // resolve through the TypeOp base defaults (typeop.cc:696/:729). ----
    {
        let target = fd.new_code_ref(Address::new(0x600160));
        let out = fd.new_varnode(8, Address::new(0x600));
        let op = fd.new_op(1, Address::new(0x500160));
        fd.op_set_opcode(&op, OpCode::CPUI_CALL);
        fd.op_set_input(&op, target, 0);
        fd.op_set_output(&op, out);
        let call = op.0.read().unwrap();
        emit_input_case("call_input0_nofspec", call_typeop, &call, 0, &type_factory);
        emit_output_case("call_output_nofspec", call_typeop, &call, &type_factory);
    }

    // ---- CALL with one locked parameter: D1 behavior must not regress ----
    {
        let mut prototype =
            FuncProto::new("typeop_localbase_fixture".to_string(), void_type.clone());
        let mut locked_ptr = ProtoParameter::new(
            "locked_ptr".to_string(),
            char_pointer.clone(),
            Address::new(0x00),
        );
        locked_ptr.flags |= protoparam_flags::TYPE_LOCKED;
        prototype.add_parameter(locked_ptr);

        let target = fd.new_code_ref(Address::new(0x600170));
        let input1 = fd.new_constant(8, 0x7180);
        let op = fd.new_op(2, Address::new(0x500170));
        fd.op_set_opcode(&op, OpCode::CPUI_CALL);
        fd.op_set_input(&op, target, 0);
        let callspec_index = fd.add_call_specs(FuncCallSpecs::new_for_op(&op, prototype));
        let callspec_owner = fd
            .get_call_specs_owner(callspec_index)
            .expect("fixture callspec owner");
        let fspec = fd.new_varnode_call_specs(&callspec_owner);
        fd.op_set_input(&op, fspec, 0);
        fd.op_set_input(&op, input1, 1);

        let call = op.0.read().unwrap();
        let actual = call_typeop.get_input_local(&call, 1);
        let repeat = call_typeop.get_input_local(&call, 1);
        emit_result("call_input1_locked", 1, 8, &actual);
        let fallback = factory_base(&type_factory, 8, TypeMetatype::Unknown);
        emit_bool(
            "case.call_input1_locked.param_identity",
            identity(&actual, &char_pointer),
        );
        emit_bool(
            "case.call_input1_locked.base_identity",
            identity(&actual, &fallback),
        );
        emit_bool(
            "case.call_input1_locked.repeat_identity",
            match (actual.as_ref(), repeat.as_ref()) {
                (Some(left), Some(right)) => Arc::ptr_eq(left, right),
                _ => false,
            },
        );
    }
}
